metsim/metsim.py

```diff
@@ -83,151 +81,198 @@ def __init__(self, params: dict):
-            p.start()
-        for p in process_handles:
-            p.join()
-
-        # Close everything out
-        self.queue.put("done")
-        io_process.join()
-
-    def run(self, met_data, locations, queue):
-        """
-        Kicks off the disaggregation and queues up data for IO
-        """
-        # Where we will store the results for writing out
-        out_dict = {}
-        method = MetSim.methods[MetSim.params.get('method', 'mtclim')]
+        nprocs = MetSim.params['nprocs']
+        self.pool = Pool(processes=nprocs)
 
-        # If output timestep is not less than daily, don't call disagg
-        disagg = int(MetSim.params['time_step']) < cnst.MIN_PER_DAY
+        # Split the input into chunks to run in parallel
+        locations = np.array_split(list(zip(self.i_idx, self.j_idx)), 
+                nprocs * cnst.CHUNK_SIZE)
 
         # Do the forcing generation and disaggregation if required
+        status = []
+        for loc_chunk in locations:
+            stat = self.pool.apply_async(
+                wrap_run,
+                args=(self.method.run, loc_chunk, self.met_data, self.disagg),
+                callback=self._unpack_results)
+            status.append(stat)
+        self.pool.close()
+
+        # Check that everything worked
+        [stat.get() for stat in status]
+        self.pool.join()
+
+    def _unpack_results(self, results):
+        """Put results into the master dataset"""
+        for result in results:
+            locs, df = result
+            i = locs['lat']
+            j = locs['lon']
+            for varname in self.params['out_vars']:
+                self.output[varname].values[:, i, j] = df[varname].values
+
+    def run(self, locations):
+        """
+        Kicks off the disaggregation and queues up data for IO
+        """
+        results = []
         for i, j in locations:
             print("Processing {} {}".format(i, j))
-            out_dict[(i, j)] = (
-                    method.run(met_data.isel(y=i, x=j).to_dataframe(),
-                               MetSim.params, disagg=disagg))
-        # TODO: fix indexing above
-        queue.put(out_dict)
+            locs = dict(lat=i, lon=j)
+            ds = self.met_data.isel(lat=i, lon=j)  # fix this indexing
+            lat = ds['lat'].values
+            elev = ds['elev'].values
+            df = ds.drop(['lat', 'lon', 'elev']).to_dataframe()
+            df = self.method.run(df, MetSim.params, elev=elev,
+                    lat=lat, disagg=self.disagg)
+            results.append((locs, df))
+        self._unpack_results(results)
 
     def find_elevation(self, lat: float, lon: float) -> float:
         """ Use the domain file to get the elevation """
-        i_idx = np.abs(self.lat - lat).argmin()
-        j_idx = np.abs(self.lon - lon).argmin()
-        return self.domain['elev'][i_idx, j_idx]
+        i = int(np.abs(self.lat - lat).argmin())
+        j = int(np.abs(self.lon - lon).argmin())
```

Comment:
Is this robust for [-180, 180> and [0,360>? (or does VIC explicitly require one of these formats already.